libdnf/rpm/repo_sack.cpp

```diff
@@ -30,4 +35,93 @@ RepoWeakPtr RepoSack::new_repo(const std::string & id, Repo::Type type) {
+        auto opt_binds = conf.opt_binds();
+        const auto & cfg_parser_sect = cfg_parser_data_iter->second;
+        for (const auto & opt : cfg_parser_sect) {
+            auto opt_binds_iter = opt_binds.find(opt.first);
+            if (opt_binds_iter != opt_binds.end()) {
+                try {
+                    auto value = opt.second;
+                    parser.substitute(value, base.get_variables());
+                    opt_binds_iter->second.new_string(Option::Priority::MAINCONFIG, value);
+                } catch (const Option::Exception & ex) {
+                    auto msg = fmt::format(
+                        R"**(Config error in section "{}" key "{}": {}: {})**",
+                        section,
+                        opt.first,
+                        ex.get_description(),
+                        ex.what());
+                    logger.warning(msg);
+                }
+            }
+        }
+    }
+}
+
+void RepoSack::new_repos_from_file(const std::string & path) {
+    auto & logger = base->get_logger();
+    ConfigParser parser;
+    parser.read(path);
+    const auto & cfg_parser_data = parser.get_data();
+    for (const auto & cfg_parser_data_iter : cfg_parser_data) {
+        const auto repo_id = cfg_parser_data_iter.first;
+        if (repo_id != "main") {
+            auto bad_char_idx = Repo::verify_id(repo_id);
+            if (bad_char_idx >= 0) {
+                logger.warning(
+                    fmt::format("Bad id for repo: {}, byte = {} {}", repo_id, repo_id[bad_char_idx], bad_char_idx));
+                continue;
+            }
+
+            auto repo = new_repo(repo_id);
+            auto repo_cfg = repo->get_config();
+            logger.debug(fmt::format(R"**(Start of loading configuration of repository "{}" from file "{}")**", repo_id, path));
+            load_config_from_parser(*repo_cfg, parser, repo_id, *base);
+            logger.debug(fmt::format(R"**(Loading configuration of repository "{}" from file "{}" done)**", repo_id, path));
+
+            if (repo_cfg->name().get_priority() == Option::Priority::DEFAULT) {
+                logger.warning(fmt::format(
+                    "Repository \"{}\" is missing name in configuration file \"{}\", using id.", repo_id, path));
+                repo_cfg->name().set(Option::Priority::REPOCONFIG, repo_id);
+            }
+        }
+    }
+}
+
+void RepoSack::new_repos_from_file() {
+    new_repos_from_file(std::filesystem::path(base->get_config().config_file_path().get_value()));
+}
+
+void RepoSack::new_repos_from_dir(const std::string & dir_path) {
+    std::filesystem::directory_iterator di(dir_path);
+    std::vector<std::filesystem::path> paths;
+    for (auto & dentry : di) {
+        auto & path = dentry.path();
+        if (dentry.is_regular_file() && path.extension() == ".repo") {
+            paths.push_back(path);
+        }
+    }
+    std::sort(paths.begin(), paths.end());
```

Comment:
Please could you add a reason why you sort paths before loading?

Reply:
Method documentation (in .hpp) says "The files in the directory are read in alphabetical order." That's the reason for sorting.
I prefer reading in alphabetical order rather than "random" (e.g. according to the age of the files).